### PR TITLE
i#1621 AArch64 CC opt: Add get_ABI_stack_alignment function.

### PR DESCRIPTION
**Pull request by fhahn**
This patch adds a get_ABI_stack_alignment function as a
follow-up of a3ff20e. It also updates prepare_for_clean_call
to check if the stack pointer is aligned properly on all
platforms.

### CODE REVIEW

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
run aarch32 tests

**Comment by fhahn** [comment]
run aarch32 tests

**Comment by fhahn** [comment]
run aarch32 tests

**Comment by fhahn** [comment]
run aarch32 tests

**Comment by fhahn** [comment]
run aarch32 tests

File: core/arch/arch.h [hunk 1]
```diff
@@ -512,6 +512,21 @@ mangle_special_registers(dcontext_t *dcontext, instrlist_t *ilist, instr_t *inst
 #endif
 void mangle_insert_clone_code(dcontext_t *dcontext, instrlist_t *ilist,
                               instr_t *instr _IF_X86_64(gencode_mode_t mode));
+/* Returns the number of bytes the stack pointer has to be aligned to. */
+static inline uint
+get_ABI_stack_alignment() {
+#ifdef X86
+ #if defined(X64) || defined(MACOS)
```
Comment (derekbruening) [inline core/arch/arch.h]: style: # in 1st column, here and several instances below

File: core/arch/arch.h [hunk 2]
```diff
@@ -512,6 +512,21 @@ mangle_special_registers(dcontext_t *dcontext, instrlist_t *ilist, instr_t *inst
 #endif
 void mangle_insert_clone_code(dcontext_t *dcontext, instrlist_t *ilist,
                               instr_t *instr _IF_X86_64(gencode_mode_t mode));
+/* Returns the number of bytes the stack pointer has to be aligned to. */
```
Comment (derekbruening) [inline core/arch/arch.h]: suggest: separate this with blank lines from the decls around it

File: core/arch/arch.h [hunk 3]
```diff
@@ -512,6 +512,21 @@ mangle_special_registers(dcontext_t *dcontext, instrlist_t *ilist, instr_t *inst
...
+static inline uint
+get_ABI_stack_alignment() {
+#ifdef X86
+ #if defined(X64) || defined(MACOS)
+    return 16;
+ #else
+    return 1;
```
Comment (derekbruening) [inline core/arch/arch.h]: This should be 4 for DR (but 16 for gcc 4.5+: see #847)

Comment (derekbruening) [inline core/arch/arch.h]: To clarify: leave at 4, but it's worth a comment with a ref to #847

**Comment by fhahn** [comment]
rerun aarch64 tests

**Comment by fhahn** [comment]
rerun aarch64 tests

**Comment by fhahn** [comment]
rerun aarch64 tests

**Comment by fhahn** [comment]
rerun aarch32 tests

File: core/arch/arch.h (hunk 2 above)
Comment (fhahn) [reply core/arch/arch.h]: done

File: core/arch/arch.h (hunk 1 above)
Comment (fhahn) [reply core/arch/arch.h]: done

**Comment by fhahn** [comment]
Coming back to this after too long! Sorry for the long absence and accidentally pushing  those additional commits. Should I create a fresh PR?

File: core/arch/arch.h [hunk 4]
```diff
@@ -512,6 +512,24 @@ mangle_special_registers(dcontext_t *dcontext, instrlist_t *ilist, instr_t *inst
 #endif
 void mangle_insert_clone_code(dcontext_t *dcontext, instrlist_t *ilist,
                               instr_t *instr _IF_X86_64(gencode_mode_t mode));
+
+/* Returns the number of bytes the stack pointer has to be aligned to. */
+static inline uint
+get_ABI_stack_alignment() {
```
Comment (derekbruening) [inline core/arch/arch.h]: style: { on new line

**Comment by fhahn** [comment]
Closed in favor of #2738, because I messed up the history, sorry about that.

**Pull request closed without merging**